File: Mediator/Mediator.php
<?php
/**
 * Created by PhpStorm.
 * User: jamesyan
 * Date: 2017/3/25
 * Time: 14:02
 */

class ChatRoom{
    public static function showMessage($user,$message){
        var_dump(date("Y-m-d H:i:s").":[{$user->getName()}]{$message}");
    }
}

class User{
    private $name;
    public function __construct($name)
    {
        $this->name=$name;
    }

    public function getName(){
        return $this->name;
    }

    public function setName($name){
        $this->name=$name;
    }

    public function sendMessage($message){
        ChatRoom::showMessage($this,$message);
    }
}


//demo
$robet=new User("Robot");
$john=new User("John");

$robet->sendMessage("Hi Join!");
$john->sendMessage("hello Robert!");

File: Mediator/Mediator.go
package main

import (
	"fmt"
	"time"
)
type ChatRoom struct {

}
func(c *ChatRoom)sendMessage(user *User,message string){
	fmt.Println(time.Now().Format("2016-01-02 15:04:05"),":[",user.getName(),"]-",message)
}
var chatRoom ChatRoom;

type User struct {
	name string
}
func(u *User)getName()string{
	return u.name;
}

func(u *User)setName(name string){
	u.name=name;
}

func(u *User)sendMessage(message string){
      chatRoom.sendMessage(u,message)
}


func main(){
	robert:=&User{"robert"}
	john:=&User{"John"}
	robert.sendMessage("hi John")
	john.sendMessage("hello robert");
}
